### PR TITLE
Move WEBGL_CONSTANTS and THREE_TO_WEBGL constants out from GLTFExporter and GLTFLoader

### PR DESCRIPTION
**Pull request by fernandojsg**
Not so happy with the name `THREEToWebGL.js` and `WebGLToTHREE` any suggestions? :)
I believe most of the mappings from `WebGLUtils` (https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLUtils.js#L13-L35 ...) could be replaced by this mappings constants as I find it more handy to use than need to create a `WebGLUtils` instance with a `renderer` and `extensions`. Maybe in another PR we could just reuse the constants mapping on the WebGLUtils function and just leave the code related to extensions there?

### CODE REVIEW

**Comment by takahirox** [comment]
The variables are glTF specific, correct?
I'm feeling like the variable names (of file names) should indicate they're for glTF.
So add prefix GLTF like `GLTF_WEBGL_CONSTANTS`?

**Comment by fernandojsg** [comment]
@takahirox these are not gltf specific, these are generic WEBGL constants (https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/Constants) that can be used in other places not just in the gltf loader.

**Comment by takahirox** [comment]
Oh, really. I didn't know that, thanks!

**Comment by fernandojsg** [comment]
Yes, the point was to avoid sending a `renderer`'s context to the exporter (As discussed in https://github.com/mrdoob/three.js/pull/11964)

**Comment by takahirox** [comment]
I see I see, that makes sense!

File: src/Three.js [hunk 1]
```diff
@@ -156,5 +156,8 @@ export { Curve } from './extras/core/Curve.js';
 export { ShapeUtils } from './extras/ShapeUtils.js';
 export { SceneUtils } from './extras/SceneUtils.js';
 export { WebGLUtils } from './renderers/webgl/WebGLUtils.js';
+export { WEBGL_CONSTANTS } from './renderers/webgl/WebGLConstants.js';
```
Comment (mrdoob) [inline src/Three.js]: I think I would keep this as:

```js
export { WebGLConstants } from './renderers/webgl/WebGLConstants.js';
```

Also, I think I would add a `fromGL()` and `toGL()` methods that does the `THREEToWebGL` and WebGLToTHREE` conversions. Just to keep everything in one file.

Comment (fernandojsg) [inline src/Three.js]: @mrdoob Do you mean including and export `fromGL`, `toGL` and `WebGLConstants` from `WebGLConstants.js`?

Comment (fernandojsg) [inline src/Three.js]: If that's the case, maybe rename the export to `constants` so you could use it as:
```js
THREE.WebGLConstants.constants.POINTS
THREE.WebGLConstants.fromGL(gl.POINTS)
THREE.WebGLConstants.toGL(THREE.FrontSide)
```

Comment (mrdoob) [inline src/Three.js]: I like this API style better.

```js
var constants = new THREE.WebGLConstants();
constants.POINTS;
constants.fromGL(gl.POINTS);
constants.toGL(THREE.FrontSide);
```

But I'm not 100%. Maybe `fromGL` and `toGL` could go on `WebGLUtils` instead.

Comment (fernandojsg) [inline src/Three.js]: I don't really like the need to create a new instance just to query a constant. I would prefer to access it directly:
```js
THREE.WebGLConstants.POINTS;
```

The same for the `fromGL` and `toGL` that I believe too that they should be part of  `WebGLUtils`
```js
THREE.WebGLUtils.fromGL(gl.POINTS)
THREE.WebGLUtils.toGL(THREE.FrontSide)
```

Comment (fernandojsg) [inline src/Three.js]: @mrdoob I've updated `WebGLUtils` and `WebGLConstants` (still WIP) but to show a possible implementation of this last option

Comment (fernandojsg) [inline src/Three.js]: Eventually `convert` and `toGL` will become the same function. In fact right now with the current modifications `convert` doesn't need a `gl` context anymore just the `extensions` dependency.

**Comment by fernandojsg** [comment]
@mrdoob I've updated the `WebGLUtils` to export `toGL` and `fromGL`. I've replaced `convert` by `toGL` in the rest of the code as it was just duplicated code and the need to create a new instance and the dependencies from a gl context and extensions are not needed anymore as we're just exposing constants that we've in the `WebGLConstants` file.
I didn't change yet the `GLTFLoader/Exporter` until you'd validate the code.

Taking a look at `constants.js`, I believe I could export the `WebGLConstants` in a similar way instead of just exporting the whole object if you prefer so.

**Comment by takahirox** [comment]
Will this PR be merged soon?
I'm making PRs which use constants.
I'll wait for this PR being merged if it'll be soon.

**Comment by fernandojsg** [comment]
Rebased!
I export `WebGLUtils` object that contains the `toGL` and `fromGL` functions that are basically just an alias to access to the mapping dictionaries.
I believe it's more intuitive and easy to use this way instead of creating a new instance of the object everytime you need to use, the same way as we don't do `var math = new Math(); math.clamp(x)` every time we need to use a math function.
@mrdoob friendly bump :wave: :angel:

**Comment by fernandojsg** [comment]
I'm not sure why the `lgtm` tests are failing they were ok I did a rebase and they seem to be broken since then ¯\_(ツ)_/¯

**Comment by takahirox** [comment]
`lgtm` tests are fine now 😄

**Comment by fernandojsg** [comment]
@takahirox yep, I just rebased it and pushed it again :)

**Comment by takahirox** [comment]
When will we merge this PR?

File: examples/js/exporters/GLTFExporter.js [hunk 2]
```diff
@@ -373,10 +336,10 @@ THREE.GLTFExporter.prototype = {
 
 			var gltfSampler = {
 
-				magFilter: THREE_TO_WEBGL[ map.magFilter ],
-				minFilter: THREE_TO_WEBGL[ map.minFilter ],
-				wrapS: THREE_TO_WEBGL[ map.wrapS ],
-				wrapT: THREE_TO_WEBGL[ map.wrapT ]
+				magFilter: THREE.WebGLUtils.toGL[ map.magFilter ],
```
Comment (mrdoob) [inline examples/js/exporters/GLTFExporter.js]: `toGL` is a function 😅

File: examples/js/loaders/GLTFLoader.js [hunk 3]
```diff
@@ -1381,7 +1264,7 @@ THREE.GLTFLoader = ( function () {
 			return parser.getDependency( 'bufferView', accessor.bufferView ).then( function ( bufferView ) {
 
 				var itemSize = WEBGL_TYPE_SIZES[ accessor.type ];
-				var TypedArray = WEBGL_COMPONENT_TYPES[ accessor.componentType ];
+				var TypedArray = THREE.WEBGL_TO_THREE[ accessor.componentType ];
```
Comment (mrdoob) [inline examples/js/loaders/GLTFLoader.js]: Wouldn't it be better to use the same approach everywhere? `THREE.WebGLUtils.fromGL()`...

File: src/renderers/WebGLRenderer.js [hunk 4]
```diff
@@ -31,7 +31,7 @@ import { Vector3 } from '../math/Vector3.js';
 import { WebGLClipping } from './webgl/WebGLClipping.js';
 import { Frustum } from '../math/Frustum.js';
 import { Vector4 } from '../math/Vector4.js';
-import { WebGLUtils } from './webgl/WebGLUtils.js';
+import { WebGLUtils } from './webgl/WebGLUtils';
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: Lets keep `WebGLUtils.js`.

File: src/renderers/webgl/WebGLUtils.js [hunk 5]
```diff
@@ -1,140 +1,212 @@
...
+
+import * as Constants from '../../constants';
+import { WebGLConstants } from './WebGLConstants';
+
+//------------------------------------------------------------------------------
+// THREE TO WEBGL
+//------------------------------------------------------------------------------
+var THREE_TO_WEBGL = {
+
+	[ Constants.NearestFilter ]: WebGLConstants.NEAREST,
```
Comment (mrdoob) [inline src/renderers/webgl/WebGLUtils.js]: Was not aware of this `[ variable ]` thing. Is it well supported?

Comment (fernandojsg) [reply src/renderers/webgl/WebGLUtils.js]: It's ES6 computed literals. I thought gulp would do the transform itself but it doesn't seems so :\

File: examples/js/exporters/GLTFExporter.js (hunk 2 above)
Comment (fernandojsg) [reply examples/js/exporters/GLTFExporter.js]: fixed 👼

File: examples/js/loaders/GLTFLoader.js (hunk 3 above)
Comment (fernandojsg) [reply examples/js/loaders/GLTFLoader.js]: fixed

File: src/renderers/WebGLRenderer.js (hunk 4 above)
Comment (fernandojsg) [reply src/renderers/WebGLRenderer.js]: fixed

File: src/renderers/webgl/WebGLUtils.js (hunk 5 above)
Comment (fernandojsg) [reply src/renderers/webgl/WebGLUtils.js]: The ugly but compatible version is to put the value of the constant as is, or try to use a gulp pluging to convert it to its literal value when creating the bundle. @mrdoob thoughts?

Comment (fernandojsg) [reply src/renderers/webgl/WebGLUtils.js]: My fault, this is es2015 and it's supported in all the browsers, but on firefox we need to add extra parenthesis: `[ (value) ]`. I'll fix it

Comment (fernandojsg) [reply src/renderers/webgl/WebGLUtils.js]: actually, on the latest version of nightly it works as expected :)

Comment (fernandojsg) [reply src/renderers/webgl/WebGLUtils.js]: @mrdoob my fault, I was testing it incorrectly on firefox O:) it's working as expected in all the browsers, also tested on mobile (android & ios).

**Comment by mrdoob** [comment]
Closing this for now.

**Pull request closed without merging**